File: agentic_harness/models/self_hosted.go
package models

import (
	"context"
	"fmt"

	"github.com/openai/openai-go/v3"
	"github.com/openai/openai-go/v3/option"

	"log/slog"
)

// This model would be hosted using llama.cpp or some other similar framework.
// This API model would use the OpenAI API spec for compatibility and thus
// and thus this is a wrapper around the OpenAI API client that implements the Model interface.
type SelfHostedModel struct {
	ctx    context.Context
	client openai.Client
}

func NewSelfHostedModel(modelBaseUri string) *SelfHostedModel {
	slog.Info("Initializing SelfHostedModel", "baseUri", modelBaseUri)
	return &SelfHostedModel{
		ctx:    context.Background(),
		client: openai.NewClient(option.WithBaseURL(modelBaseUri)),
	}
}

func (m *SelfHostedModel) Execute(messages []Message) (string, error) {

	// Convert our internal Message format to the OpenAI API format
	var apiMessages []openai.ChatCompletionMessageParamUnion
	for _, msg := range messages {
		switch msg.Role {
		case "user":
			slog.InfoContext(m.ctx, "Processing user message", "content", msg.Content)
			apiMessages = append(apiMessages, openai.UserMessage(msg.Content))
		case "system":
			slog.InfoContext(m.ctx, "Processing system message", "content", msg.Content)
			apiMessages = append(apiMessages, openai.DeveloperMessage(msg.Content))
		case "assistant":
			slog.InfoContext(m.ctx, "Processing assistant message", "content", msg.Content)
			apiMessages = append(apiMessages, openai.AssistantMessage(msg.Content))
		default:
			slog.InfoContext(m.ctx, "Processing unknown message role", "role", msg.Role)
			apiMessages = append(apiMessages, openai.UserMessage(msg.Content)) // Default to user message
		}
	}

	// Call the OpenAI API to get a response
	chatCompletion, err := m.client.Chat.Completions.New(m.ctx, openai.ChatCompletionNewParams{
		Messages: apiMessages,
	})
	if err != nil {
		return "", err
	}

	// Log the choices returned by the API for debugging
	for i, choice := range chatCompletion.Choices {
		slog.InfoContext(m.ctx, "Choice", "index", i, "content", choice.Message.Content)
	}

	if len(chatCompletion.Choices) == 0 {
		return "", fmt.Errorf("no choices returned from model")
	}

	slog.InfoContext(m.ctx, "Selected Choice", "content", chatCompletion.Choices[0].Message.Content)
	// Return the content of the first choice from the response
	return chatCompletion.Choices[0].Message.Content, nil
}

File: agentic_harness/src/models/self_hosted.ts
import OpenAI from "openai";
import { Model, modelRegistry } from "./models";
import { logger } from "@/logger";
import { ModelInput, ModelInteraction, ModelMessageOutput } from "./conversation";
import { ChatCompletionAssistantMessageParam } from "openai/resources";

function toChatCompletionInteraction(message: ModelInteraction[]): OpenAI.Chat.Completions.ChatCompletionMessageParam[] {
    return message.map((msg) => {
        if (msg.type === "message") {
            return {
                role: msg.role,
                content: msg.content,
            };
        } else if (msg.type === "tool_call") {
            return {
                role: "assistant",
                tool_calls: [{
                    id: msg.id,
                    type: "function",
                    function: {
                        arguments: JSON.stringify(msg.arguments),
                        name: msg.tool.name
                    },
                }],
            } as ChatCompletionAssistantMessageParam;
        } else if (msg.type === "tool_response") {
            return {
                role: "tool",
                content: JSON.stringify(msg.result),
                tool_call_id: msg.id,
                name: msg.tool.name,
            };
        } else {
            throw new Error(`Unknown message type: ${msg}`);
        }
    })
}

export class SelfHostedModel implements Model {

    private client: OpenAI
    private modelName: string

    constructor(baseURL: string, modelName: string, apiKey?: string) {
        this.modelName = modelName;
        this.client = new OpenAI({
            baseURL,
            apiKey: apiKey ?? "local-api-key",
        });
    }

    async execute(input: ModelInput): Promise<ModelMessageOutput[]> {

        const response = await this.client.chat.completions.create({
            model: this.modelName,
            messages: [
                {
                    role: "system",
                    content: input.systemPrompt ?? "You are a helpful assistant."
                },
                ...toChatCompletionInteraction(input.history)
            ],
            tools: input.tools.map((tool) => ({
                type: "function",
                function: {
                    name: tool.name,
                    description: tool.description,
                    parameters: tool.inputSchema
                }
            })),

        })

        logger.info(`Total tokens: ${response.usage?.total_tokens}`);

        response.choices.forEach((choice, index) => {
            logger.info(`Choice ${index + 1}: ${choice.message?.content}. Finish reason: ${choice.finish_reason}`);
        });

        if (response.choices.length === 0) {
            throw new Error("No response choices received from self-hosted model.");
        }

        const message = response.choices[0]!.message;

        const outputs: ModelMessageOutput[] = [];

        if (message.content) {
            outputs.push({
                role: "assistant",
                type: "message",
                content: message.content
            });
        }

        if (message.tool_calls && message.tool_calls.length > 0) {
            for (const tool of message.tool_calls) {
                if (tool.type !== "function")
                    continue;

                const toolToUse = input.tools.find(t => t.name === tool.function.name);

                if (!toolToUse) {
                    throw new Error(`Tool not found for function call: ${tool.function.name}`);
                }

                outputs.push({
                    type: "tool_call",
                    arguments: JSON.parse(tool.function.arguments),
                    id: tool.id,
                    tool: toolToUse
                });
            }
        }

        return outputs;
    }
}

// Self hosted models are registered with the model registry, but only if the environment variable is set
modelRegistry.registerModels((env) => {

    // environment variable prefixes
    const SELF_HOSTED_MODEL_BASE_URL_PREFIX = "SELF_HOSTED_MODEL_BASE_URL_";
    const SELF_HOSTED_MODEL_API_KEY_PREFIX = "SELF_HOSTED_MODEL_API_KEY_";

    const models = new Map<string, SelfHostedModel>();

    for (const key in env) {
        if (key.startsWith(SELF_HOSTED_MODEL_BASE_URL_PREFIX) && env[key]) {

            // get the model name from the environment variable
            const modelName = key.substring(SELF_HOSTED_MODEL_BASE_URL_PREFIX.length).toLowerCase();

            if (modelName.length === 0) {
                logger.warn(`Environment variable ${key} is missing a model name suffix, skipping registration of this self hosted model.`);
                continue;
            }

            const baseURL = env[key];

            if (!baseURL || !URL.canParse(baseURL)) {
                logger.warn(`Invalid base URL for self-hosted model: ${modelName}. Skipping registration.`);
                continue;
            }

            // look for a corresponding API key environment variable
            const apiKeyEnvVar = SELF_HOSTED_MODEL_API_KEY_PREFIX + modelName.toUpperCase();
            const apiKey = env[apiKeyEnvVar];

            // it doesn't matter if we get an api-key, api keys are optional

            // log the registration of the self-hosted model, but don't log the actual API key for security reasons
            logger.info(`Registering self-hosted model: ${modelName} with base URL: ${baseURL} and API key: ${apiKey ? "provided" : "not provided"}`);
            models.set(modelName, new SelfHostedModel(baseURL, modelName, apiKey));
        }
    }

    return models;
})